### ISSUE TITLE
RPMB question in RPI3. 

### ISSUE THREAD

**Issue by SangsubLim** [OP-TEE/optee_os#4847]
Hello,

I am studying RPMB operation using RPI3 B.
After analyzing through QEMU+RPMB_EMU, I wanted to test it on a real board.
For reference, the eMMC module(RasPiKey 16GB) I am testing is as follows.
https://www.uugear.com/product/raspikey-plug-and-play-emmc-module-for-raspberry-pi/

It uses eMMC manufactured by Samsung.
http://www.uugear.com/doc/datasheet/KLMAG1JETD-B041.pdf
When I checked, it said that a 4MB partition was provided and the eMMC 5.1 spec was supported.

I downloaded rpi3 optee and built the options below.

> make -j10 CFG_REE_FS=y CFG_RPMB_FS=y **_CFG_RPMB_WRITE_KEY=n_** CFG_RPMB_FS_DEBUG_DATA=y CFG_TEE_TA_LOG_LEVEL=3 CFG_TEE_CLIENT_LOG_LEVEL=3 CFG_TEE_SUPP_LOG_LEVEL=3 CFG_FTRACE_SUPPORT=y

And by partially modifying the CMakefile, RPMB_EMU is disabled.

The build and boot went well. In the boot log, information about RPMB was also output as shown below.

> [    3.420553] mmcblk0boot0: mmc0:0001 AJTD4R partition 1 4.00 MiB
> [    3.426644] Indeed it is in host mode hprt0 = 00001101
> [    3.437156] mmcblk0boot1: mmc0:0001 AJTD4R partition 2 4.00 MiB
> [    3.452511] mmcblk0rpmb: mmc0:0001 AJTD4R partition 3 4.00 MiB, chardev (244:0)

When I first tested xtest 6001, there was a permission problem in iocmd.
I thought this was a problem caused by tee-supplicant running as tee rather than root.
To this end, I deleted the part that executes tee-supplicant as a daemon during the booting process, and after booting, run it as tee-supplicant & in the root shell.

After that, I first ran xtest 6001, and the log is output as shown below. 

> D/TC:? 0 tee_ta_init_pseudo_ta_session:299 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
> D/TC:? 0 ldelf_load_ldelf:91 ldelf load address 0x40006000
> D/LD:  ldelf:134 Loading TS 8aaaf200-2450-11e4-abe2-0002a5d5c51b
> D/TC:? 0 ldelf_syscall_open_bin:142 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (Secure Storage TA)
> D/TC:? 0 tee_rpmb_init:1114 RPMB: Syncing device information
> D/TC:? 0 tee_rpmb_get_dev_info:927 Dumping dev_info:
> D/TC:? 0 tee_rpmb_get_dev_info:928 000000001019b980  15 01 00 41 4a 54 44 34  52 06 30 e6 58 62 c7 db
> D/TC:? 0 tee_rpmb_get_dev_info:928 000000001019b990  20 01 00
> **D/TC:? 0 tee_rpmb_init:1122 RPMB: RPMB size is 32*128 KB
> D/TC:? 0 tee_rpmb_init:1123 RPMB: Reliable Write Sector Count is 1**
> D/TC:? 0 tee_rpmb_init:1150 RPMB INIT: Deriving key
> I/TC: RPMB: Using generated key
> D/TC:? 0 tee_rpmb_init:1165 RPMB INIT: Verifying Key
> D/TC:? 0 tee_rpmb_req_pack:637 Dumping data frame 0:
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ec54  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ec64  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ec74  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ec84  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ec94  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017eca4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ecb4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ecc4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ecd4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ece4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ecf4  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed04  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed14  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed24  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed34  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed44  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed54  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed64  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed74  10 76 66 31 96 5c d4 a1  b3 a2 20 f9 a2 6e 80 f2
> D/TC:? 0 tee_rpmb_req_pack:638 000000001017ed84  00 00 00 00 00 00 00 00  00 00 00 02
> **[   35.320617] sdhost-bcm2835 3f202000.mmc: Card stuck in wrong state! card_busy_detect status: 0xd00**
> [   35.329837] sdhost_busy_irq: intmask 440, data 0000000000000000
> [   45.537450] mmc0: timeout waiting for hardware interrupt.
> [   45.542952] [43170f1a] FCM> 10363888 0
> [   45.546767] [43170f1a] CMD  d 0
> [   45.549963] [43170f1b] REQ> 10363888 0
> [   45.553776] [43170f40] TSK< 10363888 0
> [   45.557590] [43170f51] TSK> 10363888 0
> [   45.561402] [43170f65] REQ< 10363888 10801
> [   45.565567] [43170f66] CMD< d 10000
> [   45.569115] [43170f68] FCM< 10363888 10363948
> [   45.573545] [43170f6b] RSP  d00 0
> [   45.576916] [43170f70] FCM> 10363888 0
> [   45.580729] [43170f70] CMD  d 0
> [   45.583923] [43170f71] REQ> 10363888 0
> [   45.587735] [43170f93] TSK< 10363888 0
> [   45.591547] [43170fa5] TSK> 10363888 0
> [   45.595361] [43170fcc] REQ< 10363888 10801
> [   45.599527] [43170fcd] CMD< d 10000
> [   45.603076] [43170fcf] FCM< 10363888 10363948
> [   45.607506] [43170fd1] RSP  d00 0
> [   45.610878] [43170fd6] FCM> 10363888 0
> [   45.614691] [43170fd7] CMD  d 0
> [   45.617886] [43170fd7] REQ> 10363888 0
> [   45.621699] [43170ffa] TSK< 10363888 0
> [   45.625513] [4317100b] TSK> 10363888 0
> [   45.629325] [4317101f] REQ< 10363888 10801
> [   45.633491] [43171020] CMD< d 10000
> [   45.637039] [43171022] FCM< 10363888 10363948
> [   45.641469] [43171024] RSP  d00 0
> [   45.644841] [4317102a] FCM> 10363888 0
> [   45.648653] [4317102a] CMD  d 0
> [   45.651848] [4317102b] REQ> 10363888 0
> 
> ........
> 
> 
> [   80.366411] sdhost_busy_irq: intmask 440, data 0000000000000000
> E/TC:? 0 tee_rpmb_verify_key_sync_counter:1013 Verify key returning 0xffff0000
> E/TC:? 0 tee_rpmb_init:1176 Verify key failed!
> E/TC:? 0 tee_rpmb_init:1177 Make sure key here matches device key
> D/TC:? 0 ldelf_syscall_open_bin:146 res=0xffff0000
> E/LD:  init_elf:437 sys_open_ta_bin(8aaaf200-2450-11e4-abe2-0002a5d5c51b)
> E/TC:? 0 ldelf_init_with_ldelf:126 ldelf failed with res: 0xffff0000
> D/TC:? 0 tee_ta_open_session:718 init session failed 0xffff0000
> [   90.593456] mmc0: timeout waiting for hardware interrupt.

Roughly, RPMB info seems to be imported through tee-supplicant. That is, open and iocmd operation seems to work normally for RPMB dev.
However, after that, there is a problem in the process of performing RPMB_MSG_TYPE_REQ_WRITE_COUNTER_VAL_READ to verify the key.

I haven't done RPMB_MSG_TYPE_REQ_AUTH_KEY_PROGRAM on eMMC yet.
Therefore, of course, eMMC will not be able to calculate hmac.

In RPMB_EMU, at this point, TEE_ERROR_ITEM_NOT_FOUND is received from tee-supplicant and the Auth Key is input through the "tee_rpmb_write_and_verify_key" function. (in tee_rpmb_init function)

But in my RPI3 it is returning TEE_ERROR_GENERIC. The tee-supplicant log at this point is as follows. 

> ERR [189] TSUP:process_one_request:645: func : 1
> ERR [189] TSUP:rpmb_process_request_unlocked:841: START :: sreq->cmd[0]
> ERR [189] TSUP:mmc_rpmb_fd:160: dev_id = 0
> ERR [189] TSUP:rpmb_data_req:682: START :: msg_type[2]
> ERR [189] TSUP:rpmb_data_req:691: Req: 1 frame(s) of type 0x0002
> ERR [189] TSUP:rpmb_data_req:692: Rsp: 1 frame(s)
> ERR [189] TSUP:rpmb_data_req:774: CALL IOCTL :: MMC_IOC_MULTI_CMD
> **ERR [189] TSUP:rpmb_data_req:776: ioctl ret=-1 errno=110, Connection timed out
> ERR [189] TSUP:rpmb_data_req:784: END :: res[ffff0000]**
> ERR [189] TSUP:rpmb_process_request_unlocked:875: END :: res[ffff0000]
> ERR [189] TSUP:process_one_request:684: END :: ret[ffff0000]

Connection timed out occurs when iocmd sends "Send request frame" and "Read response frames" for message (RPMB_MSG_TYPE_REQ_WRITE_COUNTER_VAL_READ) processing.

I checked the document(Embedded Multi-Media Card (e•MMC) Electrical Standard (5.1A)), and I wonder if the following are related.

> **7.4.69 PARTITION_CONFIG (before BOOT_CONFIG) [179]
> Bit[2:0] : PARTITION_ACCESS (before BOOT_PARTITION_ACCESS, R/W/E_P)
> User selects partitions to access
> 0x3 : R/W Replay Protected Memory Block (RPMB)**

I'd appreciate it if you could let me know if there are any more special things that need to be done to use RPMB.

**Comment by jforissier** [comment]
Hi @SangsubLim,

Thanks for the detailed report.

> I thought this was a problem caused by tee-supplicant running as tee rather than root.
> To this end, I deleted the part that executes tee-supplicant as a daemon during the booting process, and after booting, run it as tee-supplicant & in the root shell.

That is fine for debugging, but changing the permission of the RPMB device would be better I think (set the group to 'tee' perhaps).

> Connection timed out occurs when iocmd sends "Send request frame" and "Read response frames" for message (RPMB_MSG_TYPE_REQ_WRITE_COUNTER_VAL_READ) processing

Could you try a similar operation using the `mmc` command? (Ubuntu: `apt install mmc-utils`)

**Comment by etienne-lms** [comment]
> That is fine for debugging, but changing the permission of the RPMB device would be better I think (set the group to 'tee' perhaps).

**tee-supplicant** also needs RAWIO system capability to access **/dev/mmcblk0** (0 here) device base space.
You can archive that by root rights on **tee-supplicant** or using `setcap`.
```shell
setcap cap_sys_rawio=pe /path/to/tee-supplicant

**Comment by etienne-lms** [comment]
Sorry, let me rephrase.

**tee-supplicant** needs to access **/dev/mmcblkX** to identify RPMB capabitlites. 
The operation requires access rights to **/dev/mmcblkX**, likely to planned to be in 'tee' group. We must play with groups for that. In the end, it depends on the distribution.
The operation also requires **tee-supplicant** has RAWIO system capability. See in [mmc block driver](https://elixir.bootlin.com/linux/v5.13/source/drivers/mmc/core/block.c#L732). Above is how it can be done from **S30optee** file in our buildroot distrib.

**Comment by SangsubLim** [comment]
@jforissier,
Thank you for your advice.
I understood that your reply was to test the eMMC module I purchased by installing mmc-utils on a linux pc (ubuntu).
I'm outside now, so I'll be able to check it out next week.

But there was a strange part. When I connected to my ubuntu PC after purchasing the eMMC module, rpmb dev such as /dev/mmcblk0rpmb was not found.
I can test it next week to see if there is anything I need to install additionally. If mmc-utils is installed, mmcblk0rpmb may be searched, hopefully.

As I mentioned earlier, I am having a problem when I read a write counter value. Therefore, I try to check the following contents using mmc util.
http://manpages.ubuntu.com/manpages/bionic/man1/mmc.1.html

> rpmb read-counter <rpmb device>
>     Counter value for the <rpmb device> will be read to stdout.


@etienne-lms,
I appreciate your advice.
I'm a beginner, so I could only understand part of what you said.
When the eMMC read op issue is resolved, I will take a closer look at your advice.

**Comment by SangsubLim** [comment]
@jforissier,
I've tried the following additional actions, but to no avail.
I connected the eMMC module (RasPiKey) to my Ubuntu PC using a reader, but RPMB was not found.
I installed Raspbian (2021-05-07-raspios-buster-armhf-lite.zip) on my rpi3 as a workaround.
When I searched dmesg and /dev/mmc*, I checked RPMB and found it.

> pi@raspberrypi:~ $ apt install mmc-utils


enter the command as below to check the csd information.

> pi@raspberrypi:~ $ sudo mmc extcsd read /dev/mmcblk0


After that, when the following command is input to read RPMB write counter value, connection time out occurs.

> pi@raspberrypi:~ $ sudo mmc rpmb read-counter /dev/mmcblk0rpmb
> RPMB ioctl failed: Connection timed out

When outputting dmsg, numerous logs are output as shown below. 

> [   61.164405] sdhost-bcm2835 3f202000.mmc: Card stuck in wrong state! card_busy_detect status: 0xd00
> [   61.164435] sdhost_busy_irq: intmask 440, data 00000000
> [   71.514372] mmc0: timeout waiting for hardware interrupt.
> [   71.514389] [c3f1e369] FCM> 815a1b70 0
> [   71.514399] [c3f1e369] CMD  d 0
> [   71.514408] [c3f1e369] REQ> 815a1b70 0
> [   71.514417] [c3f1e376] TSK< 815a1b70 0


In this state, when extcsd read is re-executed, it does not work.

Are there any additional things I need to do?
I inquired with the manufacturer about whether the eMMC module is bad or if RPMB is not supported.

When I checked my RPMB code in OPTEE, I saw the following message.

> /*
> * Black magic: tested on a HiKey board with a HardKernel eMMC
> * module. When postsleep values are zero, the kernel logs
> * random errors: "mmc_blk_ioctl_cmd: Card Status=0x00000E00"
> * and ioctl() fails.
> */


I am trying to purchase an eMMC module sold by HardKernel, could you tell me which module you have tested?
And if there is an eMMC module that works with RPMB in RPI3 B, I would appreciate it if you let me know.

**Comment by jforissier** [comment]
It looks like the kernel cannot access the RPMB partition, I don't know if it is a hardware or a driver issue. I have not tried to use a real device for a long time and never with a RPi... My experiments were done with a HiKey board and this eMMC module: https://www.hardkernel.com/shop/8gb-emmc-module-xu4-linux/ (PCB revision was 0.3 at the time).

**Comment by SangsubLim** [comment]
@jforissier,
thank you. I have ordered hardkernel's emmc.
However, looking at the hardkernel wiki, they say that they do not know much about RPMB and have not tested it.
As you advised, I'm going to do a bit more analysis of the linux kernel code.
